### PR TITLE
base58: Improve DecodeBase58 performance.

### PR DESCRIPTION
**Pull request by wjx**
Improve DecodeBase58 performance the same way as commit 3252208 did
for EncodeBase58.

bench_bitcoin result for Base58Decode before and after the patch, each tested three times:
Before the patch:
# Benchmark,count,min,max,average

Base58Decode,524288,0.000000976440788,0.000002053217031,0.000001960330792
Base58Decode,524288,0.000000979518518,0.000001978729415,0.000001956941560
Base58Decode,524288,0.000000986496161,0.000001988708391,0.000001964757757

After the patch:
# Benchmark,count,min,max,average

Base58Decode,786432,0.000000660018486,0.000001336600690,0.000001329504509
Base58Decode,786432,0.000000668867870,0.000001348496880,0.000001338565805
Base58Decode,786432,0.000000656273187,0.000001334869012,0.000001324702074

That's about 32% decrease for the average time.

### CODE REVIEW

**Comment by fanquake** [comment]
@wjx Can you provide any numbers as to the performance improvement? Is it inline with what was seen in #7656?

File: src/base58.cpp [hunk 1]
```diff
@@ -39,12 +41,14 @@ bool DecodeBase58(const char* psz, std::vector<unsigned char>& vch)
             return false;
         // Apply "b256 = b256 * 58 + ch".
         int carry = ch - pszBase58;
-        for (std::vector<unsigned char>::reverse_iterator it = b256.rbegin(); it != b256.rend(); it++) {
+        int i = 0;
+        for (std::vector<unsigned char>::reverse_iterator it = b256.rbegin(); (carry != 0 || i < length) && (it != b256.rend()); it++, i++) {
```
Comment (fanquake) [inline src/base58.cpp]: [Comments](https://github.com/bitcoin/bitcoin/pull/7656/files#r56476755) from the original PR suggest you should use `++i` here.

Comment (wjx) [inline src/base58.cpp]: Updated, and benchmark number attached.

**Comment by laanwj** [comment]
code review ACK https://github.com/bitcoin/bitcoin/pull/8736/commits/e892dc1268d33bfba01f35b70359b286bb9d6376

**Comment by jonasschnelli** [comment]
Code Review utACK e892dc1268d33bfba01f35b70359b286bb9d6376

**Comment by sipa** [comment]
utACK e892dc1268d33bfba01f35b70359b286bb9d6376

**Comment by fanquake** [comment]
Benchmarks
#### Before
Base58Decode,786432,0.000000720581738,0.000001479576895,0.000001459018677
Base58Decode,786432,0.000000731757609,0.000001487409463,0.000001467602791
Base58Decode,786432,0.000000714662747,0.000001470549250,0.000001451156398
Base58Decode,786432,0.000000723695848,0.000001477301339,0.000001463891143
#### After
Base58Decode,1048576,0.000000478641596,0.000001010337655,0.000000991023853
Base58Decode,1048576,0.000000484878910,0.000000999572876,0.000000988008424
Base58Decode,1048576,0.000000494663254,0.000001009095286,0.000000993179128
Base58Decode,1048576,0.000000960510079,0.000000998141331,0.000000973840770

File: src/base58.cpp [hunk 2]
```diff
@@ -25,12 +25,14 @@ bool DecodeBase58(const char* psz, std::vector<unsigned char>& vch)
...
     // Skip and count leading '1's.
     int zeroes = 0;
+    int length = 0;
     while (*psz == '1') {
         zeroes++;
         psz++;
     }
     // Allocate enough space in big-endian base256 representation.
-    std::vector<unsigned char> b256(strlen(psz) * 733 / 1000 + 1); // log(58) / log(256), rounded up.
+    int size = strlen(psz) * 733 /1000 + 1; // log(58) / log(256), rounded up.
```
Comment (paveljanik) [inline src/base58.cpp]: SPC before 1000

**Comment by paveljanik** [comment]
Confirming the numbers:
```
$ pwd; while :; do ./bench_bitcoin | grep -m1 Base58Decode; done
/tmp/bitcoin-master/src/bench
Base58Decode,786432,0.000000705200364,0.000001495651304,0.000001435880222
Base58Decode,786432,0.000000754334906,0.000001533042450,0.000001507242814
Base58Decode,786432,0.000000738342351,0.000001584410711,0.000001496629011
$ pwd; while :; do ./bench_bitcoin | grep -m1 Base58Decode; done
/tmp/bitcoin-8736/src/bench
Base58Decode,1310720,0.000000458023351,0.000000963798811,0.000000941479630
Base58Decode,1048576,0.000000513755367,0.000001057205736,0.000000997559710
Base58Decode,1048576,0.000000455460395,0.000001009773769,0.000000967173719
^C
```

**Comment by paveljanik** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/8736/commits/e892dc1268d33bfba01f35b70359b286bb9d6376

**Pull request merged**